README.md

```diff
@@ -3,3 +3,177 @@
+  }
+
+  render () {
+    const {name} = this.props.results.respository;
+    const {states: {repository: {state}}} = this.props;
+
+    // render only if data retrieved successfully
+    return state === QuerierState.Success ? (
+      <div>
+        <h1>Respository name: {name}</h1>
+        <button onClick={this.handleRepositoryStar}>Star repo</button>
+      </div>
+    ) : null;
+  }
+}
+
+// Define RepositoryDetails data requirements(queries)
+const repositoryDetailsQueries = {
+  inputQueries: {
+    repository: {
+      query: getRepository
+    }
+  },
+  actionQueries: {
+    starRepository: {
+      query: starReposoitory
+    }
+  }
+};
+
+// Decorate your component using withData HOC
+export default withData(repositoryDetailsQueries)(RepositoryDetails);
+```
+
+## Queries
+Querier gives you ability to get data either *in* and *out* of the React component. To supply data into component you use `inputQueries`. To perform async operations (e.g. submittng a form) you use `actionQueries`.
+
+### Input queries
+Here's an example of a simple input query:
+
+```js
+export const getRepository = async ({ id }: { id: number }): Promise<RepositoryType> => {
+    return await octokit.repos.getById({
+      id: id.toString()
+    }).data
+  };
+```
+
+To make `getRepository` query supply data to your component, you need to decorate your component using `withData` HOC:
+```ts
+const repositoryDetailsQueries = {
+  inputQueries: {
+    /**
+     * You declare that your component will use getRepository query.
+     * Query result will be available in props as this.props.results.repository
+     **/
+    repository: {
+      query: getRepository
+    }
+  }
+}
+
+export default withData(repositoryDetailsQueries)(YourComponent);
+```
+Input queries are fired when component mounts. **They always receive component's full set of props** as argument. Later, they are being reinvoked when component's props change.
+
+Input queriy results are **cached** by default. Please see [Caching]() for more details.
+
+### Action queries
+Action queries are very similiar to input queries. The main difference is that thay are not being invoked automatically. The purpose of action query is to perform async operation as a result of user interaction, e.g. submitting a form.
+
+Here's an example of action query usage:
+```tsx
+
+// queries.ts
+export const searchQuery = async (userName: string): Promise<SearchOrgResults> => {
+    const results = await octokit.search.users({
+      q: userName,
+    });
+
+    return results.data;
+  };
+
+// SearchComponent.tsx
+type SearchComponentActionQueries = {
+  searchUser: SearchOrgResults;
+}
+
+class SearchComponent extends React.Component<WithDataProps<{}, {} , SearchComponentActionQueries>> {
+  handleSubmit(userName: string) {
+    if (this.props.actionQueries && this.props.actionQueries.searchUser) {
+      this.props.actionQueries.searchUser(userName);
+    }
+  }
+
+  render () {
+    // ...
```

Comment:
Sample of using actionQuery state?

Reply:
This is still on TODO list. Gonna write about it soon.

Reply:
@konrad147 ready in 7f2c1eee5b7081eb6cac1dfc6071ed439efa8c69